include/rmm/detail/dynamic_load_runtime.hpp

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include <dlfcn.h>
+#include <memory>
+
+namespace rmm::detail {
+
+/**
+ * @brief `dynamic_load_runtime` loads the cuda runtime library at runtime
+ *
+ * By loading the cudart library at runtime we can use functions that
+ * are added in newer minor versions of the cuda runtime.
+ */
+struct dynamic_load_runtime {
+  static constexpr auto dlclose_destructor = [](void* handle) { ::dlclose(handle); };
+  inline static std::unique_ptr<void, decltype(dlclose_destructor)> cuda_runtime_lib{
+    nullptr, dlclose_destructor};
+
+  static bool open_cuda_runtime()
+  {
+    if (!cuda_runtime_lib) {
+      ::dlerror();
+      const int major               = CUDART_VERSION / 1000;
+      const std::string libname_ver = "libcudart.so." + std::to_string(major) + ".0";
+      const std::string libname     = "libcudart.so";
+
+      auto ptr = ::dlopen(libname_ver.c_str(), RTLD_LAZY);
+      if (!ptr) { ::dlopen(libname.c_str(), RTLD_LAZY); }
+      if (!ptr) { return false; }
+
+      cuda_runtime_lib.reset(ptr);
+    }
+    return true;
+  }
```

Comment:
A function local static saves you from needing to check if cudart was already opened and gives you thread safety (which this doesn't currently have).


```suggestion
  static void* open_cuda_runtime()
  {
    struct dlclose_deleter{ void operator()(void* handle){ ::dlclose(handle); };
    
    static std::unique_ptr<void, dlclose_deleter> cudart_handle = [](){
      ::dlerror();
      const int major               = CUDART_VERSION / 1000;
      const std::string libname_ver = "libcudart.so." + std::to_string(major) + ".0";
      const std::string libname     = "libcudart.so";
      
      auto ptr = ::dlopen(libname_ver.c_str(), RTLD_LAZY);
      if (!ptr) { ::dlopen(libname.c_str(), RTLD_LAZY); }
      if (ptr) { return ptr; }
      
      RMM_FAIL("Unable to dlopen cudart");
    }
    
    return cudart_handle.get();
  }
```

Reply:
This is really nice. I have updated the PR with a tweaked version which I found easier to read.